### PR TITLE
[PluggableDevice] Enable DefaultDevice for ops who only have host code.

### PR DESCRIPTION
**Pull request by quintinwang5**
Enable some ops with DefaultDevice for PluggableDevice.

### CODE REVIEW

**Comment by quintinwang5** [comment]
@penpornk @annarev  This PR is for some ops with DefaultDevice in plugin, please help to review. Thanks.

**Review by annarev — APPROVED** [review]
This change looks good to me. The only question I have is whether it makes sense to have both `DEVICE_CPU` and `DEVICE_DEFAULT` registrations or whether these should be merged into one. @sanjoy, @gunan in case you know preference in this case.

**Comment by quintinwang5** [comment]
> This change looks good to me. The only question I have is whether it makes sense to have both `DEVICE_CPU` and `DEVICE_DEFAULT` registrations or whether these should be merged into one. @sanjoy, @gunan in case you know preference in this case.

Seems we can't merge `DEVICE_CPU` and `DEVICE_DEFAULT` into one. Because of the colocation mechanism, some unittests will fail due to lack of colocated operators.

**Comment by jzhoulon** [comment]
@penpornk This PR is about register kernels with default device in Core TensorFlow. could you also help to review it? Thanks very much.

**Comment by gbaned** [comment]
@quintinwang5 Can you please resolve conflicts? Thanks!

**Comment by quintinwang5** [comment]
@gbaned I've resolved conflicts. Please review.

**Comment by jzhoulon** [comment]
@penpornk @allenlavoie  can you have a review on this PR? some ops(constant op) as `DEVICE_DEFAULT` may help inference performance .Thanks.

**Comment by penpornk** [comment]
@jzhoulon Sorry for the delay. I'll try to get to this PR today.

**Comment by kulinseth** [comment]
> @jzhoulon Sorry for the delay. I'll try to get to this PR today.

Thanks @penpornk and @jzhoulon!!

**Review by penpornk — APPROVED** [review]
Thank you for the PR! The kernel registrations look fine to me, although we probably should look into refactoring them at some point (most are similar to `DEVICE_GPU`).

File: tensorflow/core/kernels/shape_ops.cc [hunk 1]
```diff
@@ -124,12 +160,46 @@ REGISTER_KERNEL_BUILDER(Name("ShapeN")
                         ShapeNOp<int64>);
 #endif  // GOOGLE_CUDA || TENSORFLOW_USE_ROCM
 
+#define REGISTER_DEFAULT_KERNEL(type)                            \
```
[the remark this answers is not in the document]
Comment (jzhoulon) [reply tensorflow/core/kernels/shape_ops.cc:163]: I think it is a good idea to make them as DEVICE_DEFAULT for those kernels use HostMemory and cpu kernels.

[the remark this answers is not in the document]
Comment (penpornk) [reply tensorflow/core/kernels/shape_ops.cc:163]: We can try. I don't know if I'll be able to get it in in time though. Setting `DEVICE_DEFAULT` for kernels can affect applications and might break some tests. I'm actually looking at some of the tests this PR breaks right now.

[the remark this answers is not in the document]
Comment (quintinwang5) [reply tensorflow/core/kernels/shape_ops.cc:163]: Maybe broken tests are caused by colocation problem by some  device type? Due to colocation, `DEVICE_DEFALUT` data types should better be a subset of all registered device types' intersection.

[the remark this answers is not in the document]
Comment (kulinseth) [reply tensorflow/core/kernels/shape_ops.cc:163]: Thanks @penpornk for taking a look. Please let us know if its feasible.

**Pull request merged**